Add a Playwright e2e test

msabramo opened the pull request:
Write a Playwright test that tests the app e2e.

## Motivation and Context
Ensure things don't get broken

## How Has This Been Tested?
```
abramowi at marcs-mbp-3 in ~/Code/OpenSource/inspector (playwright-test●)
$ npx playwright test client/e2e/transport-type-dropdown.spec.ts

Running 1 test using 1 worker

  ✓  1 … Type Dropdown › should have options for STDIO, SSE, and Streamable HTTP (764ms)

  1 passed (1.5s)
```

## Breaking Changes
No

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [ ] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [ ] My code follows the repository's style guidelines
- [ ] New and existing tests pass locally
- [ ] I have added appropriate error handling
- [ ] I have added or updated documentation as needed

## Additional context
<!-- Add any other context, implementation notes, or design decisions -->

msabramo commented:
Related PR (which has perhaps fallen into some disrepair): https://github.com/modelcontextprotocol/inspector/pull/354

cliffhall commented on package.json:
```diff
@@ -37,7 +37,8 @@
     "prepare": "npm run build",
-    "publish-all": "npm publish --workspaces --access public && npm publish --access public"
+    "publish-all": "npm publish --workspaces --access public && npm publish --access public",
+    "test:e2e": "npm run test:e2e --workspace=client"
```
Please move this one up the list with the other two test scripts.

msabramo replied:
Okay, done in a2c97a0.

msabramo commented:
@olaservo: I had been thinking that this would mainly be run in CI but I like to do things in baby steps so that I don't get overwhelmed and so reviewing is easier. My plan was to do the CI in a separate PR.

Though now I'm wondering if we should just go for it since @cliffhall points out that it's not trivial to set up the tests locally and a lot of people will simply not bother, so maybe it's not of much use if it's not also run in CI.

I probably should look at your PR and see what good stuff I can borrow re: config, timeouts, etc.

msabramo commented:
Added `.github/workflows/e2e_tests.yml` (from @cliffhall) in 8a8212c9ca420f998bfa9da3f3bde2e11f23bcbb. I guess a maintainer needs to approve the workflow to run so we can see how it works?

msabramo commented:
I added support for https://github.com/daun/playwright-report-summary in cdb9180.

As a result, at https://github.com/modelcontextprotocol/inspector/actions/runs/15509171521?pr=460 can see a nice Playwright test report:

![Screenshot 2025-06-07 at 8 29 44 AM](https://github.com/user-attachments/assets/4c949d25-a5d1-4ea3-9f3d-4aff52976d5d)

msabramo commented:
@olaservo @cliffhall

olaservo commented on client/playwright.config.ts:
```diff
@@ -0,0 +1,86 @@
+        ]
+      : []),
+
+    /* Test against mobile viewports. */
```
I think if we don't want to run these lets just remove the commented parts.

olaservo reviewed (changes requested):
Looks good to me, just had a minor suggestion.

cliffhall commented:
Failures when I run the tests locally with `/usr/local/bin/npm run test:e2e`. What am I missing?

<img width="1143" alt="Screenshot 2025-06-17 at 3 20 15 PM" src="https://github.com/user-attachments/assets/dd554be9-189d-4bf2-8a8c-def78f4950f3" />

cliffhall commented:
@msabramo Also, I noticed that running these tests created a lot of files locally in the project. Can these not be created in the system tmp folder? 

I realize the CLI tests do such a thing as well, and I'd like to make that go away too, but that's for another PR.

msabramo commented:
> @msabramo Failures when I run the tests locally with `/usr/local/bin/npm run test:e2e`. What am I missing?

Well, it could probably be a few things, but one possibility is that perhaps the dev server wasn't running.

Regardless of whether that's the problem you're hitting, it seems like an easy problem to hit. I discovered that Playwright can start the server for us with a `webServer` config in `playwright.config.ts` which I added in 8b1faf1. And then I removed a similar thing in the `e2e_tests.yml` GH Actions workflow, because Playwright should take care of it inside and outside of CI, which seems better.

Give it another try with the new commits and see if it works any better.

If it still doesn't work, you might try opening a browser to http://localhost:6274 while the Playwright tests are running and opening a JavaScript console and looking for errors. I actually ran into another problem that confused me for a while where the server was starting but the web page was blank and it was because of a `SyntaxError` in the JS Console, which was resolved by running `npm install` to get some new package(s) that are necessary.

cliffhall commented:
> > @msabramo Failures when I run the tests locally with `/usr/local/bin/npm run test:e2e`. What am I missing?

> Well, it could probably be a few things, but one possibility is that perhaps the dev server wasn't running.

> Regardless of whether that's the problem you're hitting, it seems like an easy problem to hit. I discovered that Playwright can start the server for us with a `webServer` config in `playwright.config.ts` which I added in 8b1faf1. And then I removed a similar thing in the `e2e_tests.yml` GH Actions workflow, because Playwright should take care of it inside and outside of CI, which seems better.

Thanks so much for getting it to auto start the server. There were no instructions to start the server, and for a local test run, it shouldn't really have to be a requirement anyhow.

> > @msabramo Also, I noticed that running these tests created a lot of files locally in the project. Can these not be created in the system tmp folder?
> > I realize the CLI tests do such a thing as well, and I'd like to make that go away too, but that's for another PR.
> 
> You mean these:
> 
> ```shell
> $ git status
> On branch playwright-test
> Your branch is up to date with 'msabramo/playwright-test'.
> 
> Untracked files:
>   (use "git add <file>..." to include in what will be committed)
>         client/playwright-report/
>         client/results.json
>         client/test-results/
> ```
> 
> The first two are reports that are explicitly specified in `playwright.config.ts`, so they seem potentially useful and I'm not sure what `client/test-results` is for, but probably something Playwright uses, so I'd be nervous about moving them. How about if I add them to `.gitignore`? Does that help with your concern?

Please forgive the bluntness, but in an already crowded project, these folders and files are just litter to me. IMHO, adding them to `.gitignore` and the `clean` script are not an acceptable solution. The _former_ means they remain in my project, visual clutter that I do not need to see. The _latter_ means I need to run another command, which also removes all the node modules and does another `npm install`, time I do not need to waste.

When running tests, I only care about whether they failed or passed, period. The following is all I need to see:

<img width="465" alt="Screenshot 2025-06-18 at 3 36 28 PM" src="https://github.com/user-attachments/assets/d5937d4e-c53d-491a-9403-d4071ce2f11c" />

If you could simply make the test:e2e script delete these files after the run, it would be perfect.

cliffhall reviewed (changes requested):
> > @msabramo Failures when I run the tests locally with `/usr/local/bin/npm run test:e2e`. What am I missing?

> Well, it could probably be a few things, but one possibility is that perhaps the dev server wasn't running.

> Regardless of whether that's the problem you're hitting, it seems like an easy problem to hit. I discovered that Playwright can start the server for us with a `webServer` config in `playwright.config.ts` which I added in 8b1faf1. And then I removed a similar thing in the `e2e_tests.yml` GH Actions workflow, because Playwright should take care of it inside and outside of CI, which seems better.

Thanks so much for getting it to auto start the server. There were no instructions to start the server, and for a local test run, it shouldn't really have to be a requirement anyhow.

> > @msabramo Also, I noticed that running these tests created a lot of files locally in the project. Can these not be created in the system tmp folder?
> > I realize the CLI tests do such a thing as well, and I'd like to make that go away too, but that's for another PR.
> 
> You mean these:
> 
> ```shell
> $ git status
> On branch playwright-test
> Your branch is up to date with 'msabramo/playwright-test'.
> 
> Untracked files:
>   (use "git add <file>..." to include in what will be committed)
>         client/playwright-report/
>         client/results.json
>         client/test-results/
> ```
> 
> The first two are reports that are explicitly specified in `playwright.config.ts`, so they seem potentially useful and I'm not sure what `client/test-results` is for, but probably something Playwright uses, so I'd be nervous about moving them. How about if I add them to `.gitignore`? Does that help with your concern?

Please forgive the bluntness, but in an already crowded project, these folders and files are just litter to me. IMO, adding them to `.gitignore` and the `clean` script are not an acceptable solution. The _former_ means they remain in my project, visual clutter that I do not need to see. The _latter_ means I need to run another command, which also removes all the node modules and does another `npm install`, time I do not need to waste.

When running tests, I only care about whether they failed or passed, period. The following is all I need to see:

<img width="465" alt="Screenshot 2025-06-18 at 3 36 28 PM" src="https://github.com/user-attachments/assets/d5937d4e-c53d-491a-9403-d4071ce2f11c" />

If you could simply make the test:e2e script delete these files after the run, it would be perfect.

msabramo commented:
@cliffhall: Thanks for the detailed comments! I think it should be better now. With the latest commits, I don't even generate `playwright-report` or `results.json` if we're not running in CI (those are needed for the GitHub Action workflow) and `test-results` is moved to `client/e2e/test-results` and it gets removed by the `test:e2e` script at the end if we're not running in CI.

So I think things should be clean but let me know if I missed anything.

cliffhall commented on .github/workflows/e2e_tests.yml:
```diff
@@ -0,0 +1,73 @@
+            client/playwright-report/
+            client/test-results/
+            client/results.json
+          retention-days: 30
```
If this is done on every push to main, we don't need to retain these files for 30 days.

cliffhall commented on .github/workflows/e2e_tests.yml:
```diff
@@ -0,0 +1,73 @@
+            client/results.json
+          retention-days: 30
+
+      - name: Publish Playwright Test Summary
+        uses: daun/playwright-report-summary@v3
+        if: always()
+        with:
+          report-file: client/results.json
+          comment-title: "🎭 Playwright E2E Test Results"
+          job-summary: true
+          icon-style: "emojis"
+          custom-info: |
+            **Test Environment:** Ubuntu Latest, Node.js 18
+            **Browsers:** Chromium, Firefox
+
+            📊 [View Detailed HTML Report](https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }}) (download artifacts)
+          test-command: "npm run test:e2e"
```
I don't know what the intention of all this is, it seems like its running the tests twice. We've already run it above in `Run Playwright tests` and here we're running it to generate the report again...?

I think what we really want is the simple [report with comment](https://github.com/marketplace/actions/playwright-report-comment) setup. There's no need to upload the reports and save/publish them. Just the results in a PR comment will do.

![image](https://github.com/user-attachments/assets/5802e2ab-1dea-4722-9d9c-38399168bada)

msabramo replied:
> I don't know what the intention of all this is, it seems like its running the tests twice. We've already run it above in `Run Playwright tests` and here we're running it to generate the report again...?
> 
> I think what we really want is the simple [report with comment](https://github.com/marketplace/actions/playwright-report-comment) setup. There's no need to upload the reports and save/publish them. Just the results in a PR comment will do.
> 
> ![image](https://private-user-images.githubusercontent.com/871933/457118721-5802e2ab-1dea-4722-9d9c-38399168bada.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTA0MzM2NDcsIm5iZiI6MTc1MDQzMzM0NywicGF0aCI6Ii84NzE5MzMvNDU3MTE4NzIxLTU4MDJlMmFiLTFkZWEtNDcyMi05ZDljLTM4Mzk5MTY4YmFkYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNjIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDYyMFQxNTI5MDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMWQ1NzUwYjc0ZGM2M2M4YWEwYzcyYjM5NGE0ODRkYzRjYjRhNTEzMjQzODM3ZTg1ZjEwOTQwOWQ3ZWU4NGU2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.cb82LIBk48UkapRu-2yPhUE0hzmohxK8wJC43ceQyZw)

So I think this is set up to do that but it's failing to post the comment because of permissions problems:

![Screenshot 2025-06-20 at 8 30 57 AM](https://github.com/user-attachments/assets/d444f55c-e8f3-4ea7-9dae-57f368d1d301)

Maybe you can have a look and see if you are able to grant permissions so that this can work and you can see if you like the way it does comments?

cliffhall reviewed (changes requested):
Let's streamline this action to just run the tests and make a comment on the PR with the results.

cliffhall commented on .github/workflows/e2e_tests.yml:
```diff
@@ -0,0 +1,73 @@
+    branches: [main]
+  pull_request:
+    branches: [main]
+
```
```suggestion

permissions:
  pull-requests: write

```

cliffhall replied:
Based on [this discussion](https://github.com/orgs/community/discussions/26644) I believe if you add this permissions block, then It will prompt maintainers to approve workflows, and it will be able to do the test when we mash that button.

cliffhall reviewed (changes requested):
Try the below suggestion.

msabramo commented:
Still got the following after applying d5b4a6e:

![Screenshot 2025-06-20 at 11 31 30 AM](https://github.com/user-attachments/assets/232ca134-88b4-44aa-9486-0ca9d1f6c5f0)

msabramo commented:
I wonder if 1ff39d1 caused something to trigger that asked a maintainer to approve execution?

cliffhall commented:
> I wonder if [1ff39f1](https://github.com/modelcontextprotocol/inspector/commit/1ff39f13f5339a1bf87101f80848bc7b69c089b4) caused something to trigger that asked a maintainer to approve execution?

Nope. That's why I asked to remove it. It did not help, but could potentially be a risk.

msabramo commented:
> > I wonder if [1ff39f1](https://github.com/modelcontextprotocol/inspector/commit/1ff39f13f5339a1bf87101f80848bc7b69c089b4) caused something to trigger that asked a maintainer to approve execution?
> 
> Nope. That's why I asked to remove it. It did not help, but could potentially be a risk.

Okay, reverted the two changes in 9ffe63e and 3e28357.

msabramo commented:
Maybe once it gets merged, the comments will work on builds on `main` (probably not on PRs in forks, which is probably a good thing).

cliffhall commented:
> Maybe once it gets merged, the comments will work on builds on `main` (probably not on PRs in forks, which is probably a good thing).

Can you also remove the permissions block

cliffhall commented:
One last question: Is there a way to get it to stop trying to make the PR comment? It's still reporting that it can't do it. 

<img width="1440" alt="Screenshot 2025-06-20 at 5 26 20 PM" src="https://github.com/user-attachments/assets/36fbd0cf-b4ec-4231-9546-8e1a8233bc1d" />

msabramo commented:
> One last question: Is there a way to get it to stop trying to make the PR comment? It's still reporting that it can't do it.

I think c747fe8 might do the trick. Let's see...

msabramo commented:
> I think [c747fe8](https://github.com/modelcontextprotocol/inspector/commit/c747fe85d8e1f0619ed761366dd66c8c30aeced4) might do the trick. Let's see...

Now it skips that step entirely for forks, which I guess is fine. 

![Screenshot 2025-06-20 at 7 36 41 PM](https://github.com/user-attachments/assets/02999756-be0d-49e4-8462-d2acb405e044)

It might be possible to have it create the report and just not comment, but not sure if it's worth it...

msabramo commented:
![IMG_6561](https://github.com/user-attachments/assets/eeba71cf-89b7-4953-a0a8-4e95f42bde35)

cliffhall reviewed (approved):
LGTM! 👍

cliffhall commented:
Thanks for hanging in there @msabramo.
